### ISSUE TITLE
Vec<MaybeUninit<u8>> transmute to Vec<u8>

### ISSUE THREAD

**Issue by kuviman** [rust-lang/rust-clippy#7706]
Lint name: `unsound_collection_transmute`

I tried this code:

```rust
unsafe fn foo(v: Vec<std::mem::MaybeUninit<u8>>) -> Vec<u8> {
    std::mem::transmute(v)
}
```

I expected to see this happen: assuming MaybeUninit has been initialized, this should be safe, since MaybeUninit should have same layout as inner type

Instead, this happened:

```
error: transmute from `std::vec::Vec<std::mem::MaybeUninit<u8>>` to `std::vec::Vec<u8>` with mismatched layout is unsound
 --> src/lib.rs:2:5
  |
2 |     std::mem::transmute(v)
  |     ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[deny(clippy::unsound_collection_transmute)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unsound_collection_transmute
```

Btw, it seems strange, but if I instead make the function generic, everything is ok:

```rust
unsafe fn foo_generic<T>(v: Vec<std::mem::MaybeUninit<T>>) -> Vec<T> {
    std::mem::transmute(v)
}

unsafe fn foo(v: Vec<std::mem::MaybeUninit<u8>>) -> Vec<u8> {
    foo_generic(v)
}
```

### Meta

Ran on [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e042fbdde01589282a8fe321f82cc418)
